tests/test_enterprise/views/test_grant_data_sharing_permissions.py

```diff
@@ -561,11 +561,13 @@ def test_post_course_specific_consent(
             user_id=self.user.id,
             enterprise_customer=enterprise_customer
         )
+        # Data sharing consent record needs to exist to check if they've already consented before
+        # Granted being false means they haven't enrolled in the course yet
         dsc = DataSharingConsentFactory(
             username=self.user.username,
             course_id=course_id,
             enterprise_customer=enterprise_customer,
-            granted=consent_provided
```

Comment:
It shouldn't be preset on the DSC record though, it should only be getting passed in as a param in the POST request.

The DSC record gets the granted value updated to True after a learner enrolls in a course, it being set to True here is what causes the `consent_record.consent_required()` check to return False and is what made it so that this ddt test case: `(False, True, '/successful_enrollment', 'course-v1:edX+DemoX+Demo_Course', str(uuid.uuid4())),` never went through the assertions under line 602. Having DSC.granted and consent_provdided being the same value from initialization makes it impossible to hit those assertions.

DSC.granted basically means whether the learner has consented and successfully enrolled into the course before. `consent_provided` is whether or not the learner consented on this request to enroll.

If they're both False then the learner doesn't get enrolled because they didn't provide consent on this request. And if they're both True then the learner doesn't get enrolled because they had already consented and been enrolled previously.

Reply:
https://github.com/edx/edx-enterprise/blob/master/consent/mixins.py#L42

Reply:
Okay now that I've looked at this in more depth I agree that `consent_provided` shouldn't be used for both "whether the user checked the box to provide DSC" and "user had previously provided consent" - which is the cause of confusion here. Passing `False` to the DSCFactory is the right thing to do since the user will [redirect](https://github.com/edx/edx-enterprise/blob/31183667f211582ffe872740c5bde141b40e720a/enterprise/views.py#L805) to `success_url` if they hit `get` in this view after previously providing consent, and the only way to hit `post` in this view is by hitting `get` first and submitting the form from the DSC page.

Reply:
It does seem there is a distinction between ddt's `consent_required` and the `granted` field on DSC, where `consent_required` probably shouldn't be used for the initial state of `granted` like Kellie mentioned, when `consent_required` is representing the expected param from the POST request, not necessarily the initial state of `granted`.

However, per Jeff's comment, if we are always creating the DSC record with `granted=False`, would we now be missing test case(s) for if/when there happens to be an existing DSC record with `granted=True` (say, for example, that field gets updated programmatically or through Django Admin somehow)?

Reply:
We would still benefit from the following ddt test case: `(False, False, '/failure_url', 'course-v1:edX+DemoX+Demo_Course', str(uuid.uuid4()))`

Reply:
I can add that ddt test case and I will also add a new set of tests for when `granted=True` as it's own set of ddt test cases so that we don't further complicate this set of tests.

Reply:
I added a lot of test cases might as well cover all of the bases and combinations